[export] Cover more cases to copy tensor conversions.

zhxchen17 opened the pull request:
Summary:
Previously we tried to convert all .to() calls to to_copy in the graph, now some user reports that other methods like .float() is not covered: https://github.com/pytorch/PiPPy/issues/1104#issuecomment-2093352734

I think fundemantally .float() should look similar to .to() in export and this diff tries to expand the coverage of the tensor conversion methods here.

Test Plan: buck run mode/opt caffe2/test:test_export -- -r float_conversion

Differential Revision: D56951634

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -2040,6 +2040,20 @@ def forward(self, x):
+    def test_float_conversion(self):
+        class Module(torch.nn.Module):
+            def forward(self, x):
+                return x.float()
```
Can we add a test case that make sure the mutation on the output is banned?

zhxchen17 replied:
we already have one test doing this.

tugsbayasgalan commented:
cc: @albanD @bdhirsh

albanD commented:
I'm not sure what exactly you're trying to achieve here, but doing this kind of long range flag is not something we want to do in general.
This kind of approach makes it easy to solve small one-off problem but usually don't scale well when you have other components that get involved and need compose with it.

albanD reviewed (changes requested):
Putting a request change to reflect my comment above.

zhxchen17 commented:
> Member

@albanD regarding the context, what we are trying to do it to automatically convert tensor conversion methods to _to_copy ops even when there is aliasing in eager. We had some previous discussion before with @tugsbayasgalan and @bdhirsh on this topic and our conclusion is to handle this at functionalization layer.

We cannot do this properly without adding a mode flag because methods like float() doesn't have arguments like `copy`.

Another approach I can try is to add `copy` argument to each pytorch methods like `to`, but I guess that a big change to the API and I'm not sure that's the right thing to do.

I need to get this to work to support Pippy's use case linked in the PR description.

Lemme know wdyt on this. Thanks!

albanD commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -218,12 +235,18 @@ def tolist(self) -> Any:
-            if len([arg for arg in args if isinstance(arg, bool)]) <= 1:
-                return super().to(*args, **{**kwargs, "copy": True})
-        return super().to(*args, **kwargs)
+    to = _conversion_method_template("to")
+    char = _conversion_method_template("char")
+    cpu = _conversion_method_template("cpu")
+    bfloat16 = _conversion_method_template("bfloat16")
+    byte = _conversion_method_template("byte")
+    cuda = _conversion_method_template("cuda")
+    double = _conversion_method_template("double")
+    float = _conversion_method_template("float")
+    bool = _conversion_method_template("bool")
+    half = _conversion_method_template("half")
+    int = _conversion_method_template("int")
+    long = _conversion_method_template("long")
```
Given that you have to override each of these manually anyway, why isn't your override here to translate them to the appropriate .to() call? namely `to(dtype=torch.float, copy=True)` or `to(device="cuda", copy=True)` for example.

zhxchen17 replied:
I looked at the doc of float, and it's indeed pretty simple: https://pytorch.org/docs/stable/generated/torch.Tensor.float.html

I didn't know how stable it would be to translate method like `.float()` to `.to()`, but I guess if that's what you prefer to have, it's worth trying.

bdhirsh replied:
hmm @albanD I think there are two (semi) independent things here, that I'd be interested in your input on:

(1) (prior to this PR) export was unhappy with the maybe-aliasing behavior of `aten.to()`. They stated that they'd prefer be able to export models with `.to()` in them, and assume that it is non-aliasing, while detecting and banning anyone who uses `.to()` + mutation in a way that would cause this assumption to be wrong (the downside is that some programs might not be exportable, but the hypothesis is that this is rare + fixable with a model tweak)

They got this to work with a small change to functionalization: override `y = x.to(...)` to make it unconditionally copy (there's a kwarg). Then, freeze the storage of `x` and `y`, to ensure that the `copy=True` transformation is semantic-preserving (and export/functionalization errors on any programs that are not semantics-preserving.

(2) (this PR): `aten.to()` is not the only "maybe-aliasing" op. In particular, there are a bunch of tensor methods like `.float()`, `.cpu()`, etc that are also maybe-aliasing, that export would also like to be always-copy.

My question was: if all of those tensor methods were implemented by calling `Tensor.to(...)`, then export would get the behavior they want for free. But they are actually implemented by bypassing the dispatcher and calling `to_impl()`, which requires export to add all of this extra handling in `FunctionalTensor` for each case.

Do you know of any good reason we can't just update those tensor methods to go through the dispatcher and call `aten.to`? (My guess was dispatch overhead, although a single extra hop through the dispatcher seems like a small price to pay).

albanD replied:
> Do you know of any good reason we can't just update those tensor methods to go through the dispatcher and call aten.to? (My guess was dispatch overhead, although a single extra hop through the dispatcher seems like a small price to pay).

Most likely dispatch overhead yes. We have been historically trying to cut down on framework overhead where we can do so easily and safely.  Most users still use eager so I don't think we want to break that too bad.

> (prior to this PR) export was unhappy with the maybe-aliasing behavior of aten.to(). 

That is due to the fact that this is pre-dispatch IR right? If we were to use post-dispatch, all of this would be gone?

> My question was: if all of those tensor methods were implemented by calling Tensor.to(...)

I don't think that will help with .contiguous() and the like?

zhxchen17 replied:
> That is due to the fact that this is pre-dispatch IR right? If we were to use post-dispatch, all of this would be gone?

In post-dispatch, we still have aliased input-output relationship being baked into the graph, and that's not what we want.

> I don't think that will help with .contiguous() and the like?

I think contiguous() is not a concern to export at least.

albanD replied:
> I think contiguous() is not a concern to export at least.

Why not? It behaves exactly the same as these functions you're fixing here right?

zhxchen17 replied:
Actually, I think you're right. We should fix .contiguous() in export as well and thanks for pointing that out. 

We do want to decouple addressing an actual use case people need in the short term and potential failures like .contiguous(), thus my previous comment.

albanD reviewed (commented):
I'm not a specialist of this functionalization code btw. So I'm not sure if it is ok to change all the to() calls to be to(copy=True).

zhxchen17 commented:
@pytorchbot rebase

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -17,6 +17,13 @@
 not_implemented_log = torch._logging.getArtifactLogger(__name__, "not_implemented")
 
 
+def _conversion_method_template(**extra_kwargs):
```
@zhxchen17 a big comment describing why this code lives in `functional_tensor.py` would be nice. Something along the lines of:

- export doesn't like ops that are maybe-aliasing (like `aten.to()`), so it hooks off of functionalization to change `.to()` under export to be always-copy, and error on programs for this this transformation will change semantics
- there are a bunch of tensor methods that have similar behavior, like .float() or .cpu(). In theory, we should get the same handling "for free" if they all were implemented by calling into `at::to()`, but they do not call into `at::to()` in eager mode for to minimize framework overhead. So instead, we are forcing them to go through `at::to()` directly in `functional_tensor.py` so they get the same treatment

zhxchen17 commented:
@pytorchbot rebase

zhxchen17 commented:
added some comments to elaborate why we are doing this way.

zhxchen17 committed:
[export] Cover more cases to copy tensor conversions. (#125628)

Summary:

Previously we tried to convert all .to() calls to to_copy in the graph, now some user reports that other methods like .float() is not covered: https://github.com/pytorch/PiPPy/issues/1104#issuecomment-2093352734

I think fundemantally .float() should look similar to .to() in export and this diff tries to expand the coverage of the tensor conversion methods here.

Test Plan: buck run mode/opt caffe2/test:test_export -- -r float_conversion

Differential Revision: D56951634

zhxchen17 commented:
@albanD @bdhirsh lemme know if you have other comments, thanks :)

albanD commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -17,6 +17,27 @@
+# relationship between input and output tensors will be baked into the graph.
+# For example, if we got a tensor with device cpu and call tensor.to("cpu"),
+# it will become a no-op in the graph. For a whole graph capture, this is not
+# sound so we need to do something different. Instead, in export we will try to
```
Wait, this banning of mutations only happens in export? But this logic is used from other places right? Which would make them unsound?

zhxchen17 replied:
> But this logic is used from other places right?

This logic is only used in export right now. We made a check in functionalization to only trigger this under export.

albanD replied:
Ok!
